pkg/client/add_events.go

```diff
@@ -163,149 +166,149 @@ func (client *DataSetClient) AddEvents(bundles []*add_events.EventBundle) error
+		}
+		if !buf.HasEvents() {
+			return buf
+		}
 		client.publishBuffer(buf)
-		return getBuffer(key)
+		return client.newBufferForEvents(buf.Session, buf.SessionInfo())
 	}
 
-	for processedMsgCnt := 0; ; processedMsgCnt++ {
-		msg, channelReceiveSuccess := <-ch
-		if !channelReceiveSuccess {
-			client.Logger.Error(
-				"Cannot receive EventWithMeta from channel",
-				zap.String("key", key),
-				zap.Any("msg", msg),
-			)
-			client.statistics.EventsBrokenAdd(1)
+	process := func(buf *buffer.Buffer, bundle EventWithMeta) *buffer.Buffer {
+		defer func() {
 			client.lastAcceptedAt.Store(time.Now().UnixNano())
-			continue
+		}()
+		if buf == nil {
+			buf = client.newBufferForEvents(key, &bundle.SessionInfo)
+		}
+		added, err := buf.AddBundle(bundle.EventBundle)
+		if err != nil {
+			client.Logger.Warn("Cannot add bundle", zap.Error(err))
 		}
 
-		bundle, ok := msg.(EventWithMeta)
-		if ok {
-			buf := getBuffer(key)
+		if buf.ShouldSendSize() || added == buffer.TooMuch && buf.HasEvents() {
+			buf = publish(key, buf)
+		}
 
-			added, err := buf.AddBundle(bundle.EventBundle)
+		if added == buffer.TooMuch {
+			added, err = buf.AddBundle(bundle.EventBundle)
 			if err != nil {
-				if errors.Is(err, &buffer.NotAcceptingError{}) {
-					buf = getBuffer(key)
-				} else {
-					client.Logger.Error("Cannot add bundle", zap.Error(err))
-					// TODO: what to do?
-					// this error happens when the buffer is already publishing
-					// we can solve it by waiting little bit
-					// since we are also returning TooMuch we will get new buffer
-					// few lines below
-					time.Sleep(10 * time.Millisecond)
-				}
+				client.Logger.Error("Cannot add bundle", zap.Error(err))
+				client.statistics.EventsDroppedAdd(1)
+				return buf
 			}
-
-			if buf.ShouldSendSize() || added == buffer.TooMuch && buf.HasEvents() {
+			if buf.ShouldSendSize() {
 				buf = publish(key, buf)
 			}
-
 			if added == buffer.TooMuch {
-				added, err = buf.AddBundle(bundle.EventBundle)
-				if err != nil {
-					if errors.Is(err, &buffer.NotAcceptingError{}) {
-						buf = getBuffer(key)
-					} else {
-						client.Logger.Error("Cannot add bundle", zap.Error(err))
-						client.statistics.EventsDroppedAdd(1)
-						continue
-					}
-				}
-				if buf.ShouldSendSize() {
+				client.Logger.Fatal("Bundle was not added for second time!", buf.ZapStats()...)
+				client.statistics.EventsDroppedAdd(1)
+				return buf
+			}
+		}
+		// TODO: Remove from adding to buffer and move to publish
+		// client.statistics.EventsProcessedAdd(1)
+		return buf
+	}
+
+	lifeTime := time.Hour
+	if client.Config.BufferSettings.MaxLifetime > 0 {
+		lifeTime = client.Config.BufferSettings.MaxLifetime
+	}
+	tickerLifetime := time.NewTicker(lifeTime)
+
+	purgeTime := time.Hour
+	if client.Config.BufferSettings.PurgeOlderThan > 0 {
+		purgeTime = client.Config.BufferSettings.PurgeOlderThan
+	}
+	tickerPurge := time.NewTicker(purgeTime)
+
+	var buf *buffer.Buffer = nil
+loopEvents:
```

Comment:
I can remove it, since `return` will work here as well. 

Based on my understanding, this is the only valid construction, how to get from the for loop inside select - https://go.dev/ref/spec#Break_statements. How would you solve it?

Fixed!